File: sumroottoleafnumbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.total = 0
        sum = 0
        
        def pre(root, sum):
            if not root:
                return
            sum = sum*10+root.val
            if not root.left and not root.right:
                self.total += sum
            pre(root.left, sum)
            pre(root.right, sum)
            
        pre(root, sum)
        return self.total
        

File: cpp/sumroottoleafnumbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int res = 0;
        string path = "";
        dfs(root, path, res);
        return res;
    }
    void dfs(TreeNode* root, string& path, int& res) {
        if (!root) return;
        path.push_back(root->val + '0');
        if (!root->left and !root->right) 
            res += stoi(path);
        dfs(root->left, path, res);
        dfs(root->right, path, res);
        path.pop_back();
    }
};
